llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -0,0 +1,152 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "CompileTimePropertiesPass.h"
+#include "DeviceGlobals.h"
+
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+
+using namespace llvm;
+
+namespace {
+
+constexpr StringRef SYCL_HOST_ACCESS_ATTR = "host_access";
+
+constexpr StringRef SPIRV_DECOR_MD_KIND = "spirv.Decorations";
+// The corresponding SPIR-V OpCode for the host_access property is documented
+// in the SPV_INTEL_global_variable_decorations design document:
+// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceGlobal/SPV_INTEL_global_variable_decorations.asciidoc#decoration
+constexpr uint32_t SPIRV_HOST_ACCESS_DECOR = 6147;
+constexpr uint32_t SPIRV_HOST_ACCESS_DEFAULT_VALUE = 2; // Read/Write
+
+enum class DecorValueTy {
+  // the value is an unsigned number (uint32_t)
+  uint32,
+  // the value is a boolean value (bool)
+  boolean,
+};
+
+struct Decor {
+  uint32_t Code;
+  DecorValueTy Type;
+};
+
+#define SYCL_COMPILE_TIME_PROPERTY(PropertyName, Decoration, ValueType)        \
+  {PropertyName, {Decoration, ValueType}},
+
+const StringMap<Decor> SpirvDecorMap = {
+#include "CompileTimeProperties.def"
+};
+#undef SYCL_COMPILE_TIME_PROPERTY
+
+/// Builds a metadata node for a SPIR-V decoration (both decoration code
+/// and value are \c uint32_t integers).
+///
+/// @param Ctx    [in] the LLVM Context.
+/// @param OpCode [in] the SPIR-V OpCode code.
+/// @param Value  [in] the SPIR-V decoration value.
+///
+/// @returns a pointer to the metadata node created for the required decoration
+/// and its value.
+MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
+                                uint32_t Value) {
+  auto *Ty = Type::getInt32Ty(Ctx);
+  SmallVector<Metadata *, 2> MD;
+  MD.push_back(ConstantAsMetadata::get(
+      Constant::getIntegerValue(Ty, APInt(32, OpCode))));
+  MD.push_back(
+      ConstantAsMetadata::get(Constant::getIntegerValue(Ty, APInt(32, Value))));
+  return MDNode::get(Ctx, MD);
+}
+
+/// Builds a metadata node for a SPIR-V decoration (both decoration code
+/// and value are \c uint32_t integers, and the secondary extra operand is a
+/// string).
+///
+/// @param Ctx        [in] the LLVM Context.
+/// @param OpCode     [in] the SPIR-V OpCode code.
+/// @param Value      [in] the SPIR-V decoration value.
+/// @param Secondary  [in] the secondary "extra operands" (\c StringRef).
+///
+/// @returns a pointer to the metadata node created for the required decoration
+/// and its value.
+MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
```

Comment:
I don't think so because in the future there would be more combinations of possible decoration extra-arguments and they will be able to have different types, not `StringRef` only. So, I prefer to leave the functions as is.

Reply:
A good idea can be to templatize the code when new types of the extra arguments appear on the horizon and calculate the required method of adding the argument (`MDString::get(...)`, `Constant::getIntegerValue(Ty, APInt(32, ...)`, etc.) in the compile time.